File: 300.500/478.GenerateRandomPointInACircle.cpp
#include <iostream>
#include <vector>
#include <random>
using namespace std;

class Solution
{
public:
    Solution(double radius, double x_center, double y_center)
        : dis(-radius, radius), xc(x_center), yc(y_center), r(radius)
    {
    }

    vector<double> randPoint()
    {
        // 在一个正方形取点
        // 并且排除在圆外的点
        while (true)
        {
            double x = dis(gen), y = dis(gen);
            // 根据勾股定理
            if (x * x + y * y <= r * r)
            {
                return {xc + x, yc + y};
            }
        }
    }

private:
    mt19937 gen{random_device{}()};
    uniform_real_distribution<double> dis;
    double xc, yc, r;
};

class Solution1
{
public:
    Solution1(double radius, double x_center, double y_center)
        : dis(0, 1), xc(x_center), yc(y_center), r(radius)
    {
    }

    vector<double> randPoint1()
    {
        double u = dis(gen);
        double theta = dis(gen) * 2 * acos(-1.0);
        double r = sqrt(u);
        return {xc + r * cos(theta) * this->r, yc + r * sin(theta) * this->r};
    }

private:
    mt19937 gen{random_device{}()};
    uniform_real_distribution<double> dis;
    double xc, yc, r;
};

void Print(vector<double> d)
{
    cout << "[" << d[0] << "," << d[1] << "]" << endl;
}

int main(int argc, char const *argv[])
{
    Solution cls(1.0, 0.0, 0.0);
    Print(cls.randPoint());
    Print(cls.randPoint());
    Print(cls.randPoint());

    Solution1 cls1(1.0, 0.0, 0.0);
    Print(cls1.randPoint1());
    Print(cls1.randPoint1());
    Print(cls1.randPoint1());
    return 0;
}

File: go/300.500/478.GenerateRandomPointInACircle.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

type Solution struct {
	radius, xCenter, yCenter float64
}

func Constructor(radius float64, x_center float64, y_center float64) Solution {
	return Solution{
		radius:  radius,
		xCenter: x_center,
		yCenter: y_center,
	}
}

func (s *Solution) RandPoint() []float64 {
	for {
		x := rand.Float64()*2 - 1
		y := rand.Float64()*2 - 1
		if x*x+y*y < 1 {
			return []float64{s.xCenter + x*s.radius, s.yCenter + y*s.radius}
		}
	}
}

func (s *Solution) RandPoint1() []float64 {
	r := math.Sqrt(rand.Float64())
	sin, cos := math.Sincos(rand.Float64() * 2 * math.Pi)
	return []float64{s.xCenter + r*cos*s.radius, s.yCenter + r*sin*s.radius}
}

func main() {
	cls := Constructor(1.0, 0.0, 0.0)
	fmt.Println(cls.RandPoint())
	fmt.Println(cls.RandPoint())
	fmt.Println(cls.RandPoint())

	fmt.Println(cls.RandPoint1())
	fmt.Println(cls.RandPoint1())
	fmt.Println(cls.RandPoint1())
}
